<a href="https://colab.research.google.com/github/mars241/Machine-Learning/blob/main/Corr_FAFI_Moncef_4_4_ML_pipeline_PCA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tuto quête

In [ ]:
import pandas as pd
import seaborn as sns
X = sns.load_dataset("iris").iloc[:,:-1]

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X)
X_scaled = scaler.transform(X)

# Dataset dimension, it's a numpy array
X_scaled.shape # >>> (150,4)

(150, 4)

In [ ]:
from sklearn.decomposition import PCA

pca = PCA(n_components=2)
X_pca = pca.fit_transform(X_scaled)

# Dataset dimension, it's a numpy array
X_pca.shape # >>> (150,2)

(150, 2)

In [ ]:
pca = PCA()
pca.fit(X_scaled)
X_pca = pca.transform(X_scaled)

# Dataset dimension, it's a numpy array
X_pca.shape # >>> (150,4)


(150, 4)

In [ ]:
pca.explained_variance_ratio_


array([0.72962445, 0.22850762, 0.03668922, 0.00517871])

In [ ]:
pca = PCA(n_components = 0.7)
pca.fit(X_scaled)
X_pca = pca.transform(X_scaled)
X_pca.shape # >>> (150,1)

(150, 1)

# Début quête

## Challenge : Réduisez les tous : ACP !


In [ ]:
import pandas as pd
import numpy as np

import plotly.express as px

from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier



lien = "https://raw.githubusercontent.com/murpi/wilddata/master/quests/weather2020.csv"
df = pd.read_csv(lien)

### Charge le dataset et mène une analyse exploratoire : le nombre de features, les corrélations dans les données, les indicateurs statistiques (de position, de dispersion)...


In [ ]:
df.shape

(366, 24)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   DATE                    366 non-null    object 
 1   MAX_TEMPERATURE_C       366 non-null    int64  
 2   MIN_TEMPERATURE_C       366 non-null    int64  
 3   WINDSPEED_MAX_KMH       366 non-null    int64  
 4   TEMPERATURE_MORNING_C   366 non-null    int64  
 5   TEMPERATURE_NOON_C      366 non-null    int64  
 6   TEMPERATURE_EVENING_C   366 non-null    int64  
 7   PRECIP_TOTAL_DAY_MM     366 non-null    float64
 8   HUMIDITY_MAX_PERCENT    366 non-null    int64  
 9   VISIBILITY_AVG_KM       366 non-null    float64
 10  PRESSURE_MAX_MB         366 non-null    int64  
 11  CLOUDCOVER_AVG_PERCENT  366 non-null    float64
 12  HEATINDEX_MAX_C         366 non-null    int64  
 13  DEWPOINT_MAX_C          366 non-null    int64  
 14  WINDTEMP_MAX_C          366 non-null    in

In [ ]:
import plotly.express as px

px.imshow(df.select_dtypes(include=['int', 'float']).corr(), text_auto=True)

In [ ]:
df.describe()

MAX_TEMPERATURE_C  MIN_TEMPERATURE_C  WINDSPEED_MAX_KMH  \
count         366.000000         366.000000         366.000000   
mean           16.997268          11.259563          20.991803   
std             6.369232           4.510670           8.546565   
min             5.000000           0.000000           3.000000   
25%            12.000000           8.000000          14.000000   
50%            16.000000          11.000000          20.000000   
75%            21.000000          15.000000          26.000000   
max            38.000000          23.000000          50.000000   

       TEMPERATURE_MORNING_C  TEMPERATURE_NOON_C  TEMPERATURE_EVENING_C  \
count             366.000000          366.000000             366.000000   
mean               10.631148           15.770492              15.275956   
std                 4.601133            6.018857               6.450046   
min                -1.000000            3.000000               3.000000   
25%                 8.000000           11.000000              10.000000   
50%                11.000000           15.000000              14.000000   
75%                14.000000           20.000000              20.000000   
max                23.000000           35.000000              37.000000   

       PRECIP_TOTAL_DAY_MM  HUMIDITY_MAX_PERCENT  VISIBILITY_AVG_KM  \
count           366.000000            366.000000         366.000000   
mean              3.169126             86.603825           9.237022   
std               6.408436              7.501151           1.059369   
min               0.000000             47.000000           3.000000   
25%               0.000000             83.000000           8.750000   
50%               0.400000             88.500000           9.875000   
75%               2.900000             92.000000          10.000000   
max              53.300000             99.000000          10.000000   

       PRESSURE_MAX_MB  ...  DEWPOINT_MAX_C  WINDTEMP_MAX_C  \
count       366.000000  ...      366.000000      366.000000   
mean       1019.562842  ...       11.256831        9.707650   
std           8.309106  ...        4.225123        5.728018   
min         982.000000  ...        0.000000       -5.000000   
25%        1016.000000  ...        8.000000        6.000000   
50%        1020.000000  ...       12.000000       10.000000   
75%        1024.000000  ...       14.000000       14.000000   
max        1044.000000  ...       22.000000       23.000000   

       WEATHER_CODE_MORNING  WEATHER_CODE_NOON  WEATHER_CODE_EVENING  \
count            366.000000         366.000000            366.000000   
mean             160.926230         166.172131            181.278689   
std               74.537276          81.604763             93.009523   
min              113.000000         113.000000            113.000000   
25%              116.000000         116.000000            116.000000   
50%              119.000000         116.000000            122.000000   
75%              176.000000         176.000000            176.000000   
max              353.000000         389.000000            389.000000   

       TOTAL_SNOW_MM    UV_INDEX     SUNHOUR       MONTH         DAY  
count          366.0  366.000000  366.000000  366.000000  366.000000  
mean             0.0    3.639344    9.269126    6.513661   15.756831  
std              0.0    1.681807    3.399074    3.455958    8.823592  
min              0.0    1.000000    3.300000    1.000000    1.000000  
25%              0.0    3.000000    6.700000    4.000000    8.000000  
50%              0.0    4.000000    9.100000    7.000000   16.000000  
75%              0.0    5.000000   11.600000    9.750000   23.000000  
max              0.0    8.000000   14.500000   12.000000   31.000000  

[8 rows x 22 columns]

In [ ]:
fig1 = px.box(df.select_dtypes(include=['int', 'float']).corr())


fig2 = px.box(df.select_dtypes(include=['int', 'float']).drop(columns=['PRESSURE_MAX_MB','WEATHER_CODE_MORNING',
       'WEATHER_CODE_NOON', 'WEATHER_CODE_EVENING']))


fig1.show()
fig2.show()

### Crée un slice du dataset avec uniquement les colonnes numériques (tu peux utiliser la méthode .select_dtypes(include='number') pour t'aider). Tu nommeras ce slice : X. Ce slice X doit avoir 366 lignes et 22 colonnes. Crée un second slice avec uniquement la colonne OPINION. Tu nommeras ce slice y.

In [ ]:
X = df.select_dtypes(include='number')
y = df['OPINION']

### Utilise un scaler pour mettre à l'échelle tes données. Tu transformeras tes données avec ce scaler, et tu enregistreras les données mises à l'échelle dans X_scaled, qui doit également avoir 366 lignes et 22 colonnes.

In [ ]:
scaler = StandardScaler()
scaler.fit(X)
X_scaled = scaler.transform(X)

X_scaled.shape


(366, 22)

### Instancie un objet pca, et entraine-le sur tes données standardisées.


In [ ]:
pca = PCA()
pca.fit(X_scaled)
X_pca = pca.transform(X_scaled)

X_pca.shape

(366, 22)

### Combien de composantes principales expliquent au moins 70% de la variance ? Tu peux t'aider de l'attribut pca.explained_variance_ratio_. Et 80% ?


In [ ]:
# 4 colonnes suffisent pour 70% de la variance

pca = PCA(n_components = 0.7)
pca.fit(X_scaled)
X_pca = pca.transform(X_scaled)
X_pca.shape

(366, 4)

In [ ]:
pca.explained_variance_ratio_

array([0.40997413, 0.20876414, 0.06448527, 0.05962773])

In [ ]:
# 4 colonnes suffisent pour 70% de la variance

pca = PCA(n_components = 0.8)
pca.fit(X_scaled)
X_pca = pca.transform(X_scaled)
X_pca.shape

(366, 6)

In [ ]:
pca.explained_variance_ratio_

array([0.40997413, 0.20876414, 0.06448527, 0.05962773, 0.04577155,
       0.03909456])

### Transforme tes données grâce à pca.fit_transform(X_scaled), et stocke le résultat dans X_pca. Affiche sur une nuage de points les 2 premières colonnes (celles qui expliquent le plus de variance). Mets une couleur différente pour chaque OPINION. Tu devrais obtenir un graphique de ce type :


In [ ]:
pca = PCA(n_components = 0.7)
pca.fit(X_scaled)
X_pca = pca.fit_transform(X_scaled)
X_pca.shape

(366, 4)

In [ ]:
px.scatter(X_pca, x = 0, y = 1, color = y)

### Lance une classification KNN sur X_scaled et y. Quel est ton score ? Lance une autre classification KNN sur les deux premières colonnes de X_pca et y. Le score est-il très différent ?

In [ ]:
modelKNN = KNeighborsClassifier()
modelKNN.fit(X_scaled, y)
modelKNN.score(X_scaled, y)

0.8661202185792349

In [ ]:
modelKNN = KNeighborsClassifier()
modelKNN.fit( X_pca, y)
modelKNN.score( X_pca, y)

0.7950819672131147

# Exercice 4 :



Selon vous quels sont les impacts de la gestion de la production d'un organisme de sécurité sociale tel que la CAF ? Quels en sont les enjeux et complexités ?

S'agissant d'une question ouverte qui arrive en fin de test, y répondre hors délai ne serait pas d'un grand intérêt.

Plutôt que de développer le propos, les points essentiels tournent autour de :  

- Distribution efficiente des prestations sociales
  - Réduction des délais de traitement
  - Amélioration continue de la qualité de réponses aux besoins des bénéficiaires.

- Optimisation des coûts :
  - Optimisation de l'utilisation des ressources
  - Automatisation des processus.  
  
- Réduction des fraudes et erreurs
  - Comme pour la détection de fraudes bancaires, piratages, probabilités d'accidents etc... des modèles de machine learning permettent la détection et le contrôle des vérifications appropriés
  - Amélioration continue avec prise en compte des feedbacks utilisateurs

- La complexité des processus :
  - diversité des prestations sociales proposées et des réglementations
  - Connaissance et expertise métier indispensable
  - Transfert de connaissances et formations dans les services pour éviter le chaos des départs/arrêts
  - Evolution et bien-être des salarié en balance avec la productivité

- Les contraintes budgétaires :
  - Trouver l'équilibre entre optimisation des coûts et garantie de la qualité des prestations.

- Communication claire :
  - Collaboration de plusieurs départements et équipes.
  - Nécessité d'avoir des process établis et respectés pour s'assurer du bon fonctionnement de toute équipe impliquée

- La sécurité des données
  - manipulation de données sensibles sur les bénéficiaires.
  - Mesures de sécurité appropriées pour protéger les données et garantir la confidentialité